drivers/gpio/gpio_hogs.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2022-2023 Vestas Wind Systems A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/init.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(gpio_hogs, CONFIG_GPIO_LOG_LEVEL);
+
+struct gpio_hog_dt_spec {
+	gpio_pin_t pin;
+	gpio_flags_t flags;
+};
+
+struct gpio_hogs {
+	const struct device *port;
+	const struct gpio_hog_dt_spec *specs;
+	size_t num_specs;
+};
+
+/* Static initializer for a struct gpio_hog_dt_spec */
+#define GPIO_HOG_DT_SPEC_GET_BY_IDX(node_id, idx)						\
+	{											\
+		.pin = DT_GPIO_HOG_PIN_BY_IDX(node_id, idx),					\
+		.flags = DT_GPIO_HOG_FLAGS_BY_IDX(node_id, idx) |				\
+			 COND_CODE_1(DT_PROP(node_id, input), (GPIO_INPUT),			\
+				     COND_CODE_1(DT_PROP(node_id, output_low),			\
+						 (GPIO_OUTPUT_INACTIVE),			\
+						 COND_CODE_1(DT_PROP(node_id, output_high),	\
+							     (GPIO_OUTPUT_ACTIVE), (0)))),	\
+	}
+
+/* Expands to 1 if node_id is a GPIO controller, 0 otherwise */
+#define GPIO_HOGS_NODE_IS_GPIO_CTLR(node_id)			\
+	DT_PROP_OR(node_id, gpio_controller, 0)
+
+/* Expands to 1 if node_id is a GPIO hog, 0 otherwise */
+#define GPIO_HOGS_NODE_IS_GPIO_HOG(node_id)			\
+	DT_PROP_OR(node_id, gpio_hog, 0)
+
+/* Expands to 1 if GPIO controller node_id has GPIO hog children, 0 otherwise */
+#define GPIO_HOGS_GPIO_CTLR_HAS_HOGS(node_id)			\
+	COND_CODE_0(						\
+		IS_EMPTY(DT_FOREACH_CHILD_STATUS_OKAY(node_id,	\
+			GPIO_HOG_NODE_IS_HOG)),			\
```

Comment:
This doesn't look right; is it?

Reply:
Ouch, good catch. This functionality was broken during a macrobatics refactoring session. It ended up just checking if the GPIO controller node had any children. Fixed now. Thank you.